client/engine/chainservice/mockchain.go

```diff
@@ -61,12 +61,15 @@ func (mc MockChain) handleTx(tx protocols.ChainTransaction) {
 		AdjudicationStatus: protocols.AdjudicationStatus{TurnNumRecord: 0},
 	}
 	for _, out := range mc.out {
-		sendEvent(out, event)
+		attemptSend(out, event)
 	}
 
 }
 
-// sendEvent sends event to the supplied chan
-func sendEvent(out chan Event, event Event) {
-	out <- event
+// attemptSend sends event to the supplied chan, and drops it if the chan is full
+func attemptSend(out chan Event, event Event) {
+	select {
+	case out <- event:
+	default:
+	}
```

Comment:
Can you explain this for a new gopher?

Is this understanding correct?
1. if `out` receives `event`, do nothing
2. else, do nothing

If `out`'s buffer is full, and it refuses to receive `event`, does `out <- event` evaluate to a "falsy" value or something?

Reply:
Your understanding is pretty much correct. With the go `select` statement, we try each case in turn until one succeeds. So if we can send `event` down the `out` chan, we're done and exit the statement. Otherwise, we revert to the default case, which really does nothing and exits the statement. 

I guess you can think of it as `out <- event` being falsy -- my mental model is that this particular pipe is blocked so you can't flow down it.

Reply:
Edit: "in turn" isn't quite right. 

> A select blocks until one of its cases can run, then it executes that case. It chooses one at random if multiple are ready.

https://go.dev/tour/concurrency/5